### ISSUE TITLE
Html parser is not robust enough

### ISSUE THREAD

**Issue by Chris00** [mirage/ocaml-cow#62]
Example:
```ocaml
# let s = "<p><img src=\"http://typeocaml.com/content/images/2014/11/thunk.jpg#hero\" alt=\"thunk\">\nA thunk is simply a function with the <em>unit</em> parameter. For example:</p>\n\n<pre><code class=\"ocaml\">let f() = 1 + 2 * 3;;  \n</code></pre>\n\n<h1 id=\"features\">Features</h1>\n\n<p>It is indeed a function and seems simple enough. However, don't overlook it. Not having any parameters makes it so special that it even has such a particular name, <em>thunk</em>. Let's have a look at its features. </p>\n\n<h2 id=\"determinedresult\">Determined result</h2>\n\n<p>If a function has parameters, the result of its application may depend on its <a href=\"http://stackoverflow.com/questions/156767/whats-the-difference-between-an-argument-and-a-parameter\">arguments</a>. For example:</p>\n\n<pre><code class=\"ocaml\">let g x y = x + y;;  \n</code></pre>\n\n<p>Without giving values for x and y, we won't be able to know what concrete value that g would give back to us.</p>\n\n<p>Thunk, however, does not have any parameters, which means its result must be fixed. After defining it, the body will not be affected by anything any more. It is like a sealed box: we may not know what's inside yet, but we know that once it is there, its content won't change. For the <code>f</code> at the beginning, we know its value will definitely be 7. For <code>let f1() = h 2 (h 3 4)</code>, we may not know the value of its body but we are sure its body is two applications of <code>h</code> and it won't change.</p>\n\n<h2 id=\"laterevaluation\">Later evaluation</h2>\n\n<p>Binding also has determined result and it is immediately evaluated after we define it. For example, <code>let x = 1 + 2 * 3</code> will return you x with 7 at once. </p>\n\n<p>Thunk is different. When we hold a thunk, we must call it like <code>f()</code> to get its result. The evaluation is a kind of delayed, but we have the full control. Here is a demonstration:</p>\n\n<pre><code class=\"ocaml\">(*Will fail immediately after you hit enter*)\nlet y = 10 / 0;;\n\n(*Will not fail but return a thunk*)\nlet f_div0() = 10 / 0;;\n\n(*Now fail*)\nf_div0();;  \n</code></pre>\n\n<h2 id=\"perfectcomputationcapsule\">Perfect computation capsule</h2>\n\n<p>With the help of the above two features, thunk encapsulates computations without evaluation, in other words, thunk does not store the actual value; instead, it stores the way of how the value would be computed. And the actual compuations will be carried out only when you decide so. Here is a case that desires thunk:</p>\n\n<pre><code class=\"ocaml\">let print_prime prime x =  \n  if x then print_int prime\n  else print_int 0;; \n\n(*Assume find_nth_prime n is already somewhere*)\nprint_prime (find_nth_prime 1000000) false;;  \n</code></pre>\n\n<p><code>print_prime</code> prints the prime only when x is true. It sounds ok. But the next function application shows a flaw that it will not print the 1,000,000th prime, but anyway the prime will be computed. What a waste of cpu cycles!</p>\n\n<p>With thunk, it gets better:  </p>\n\n<pre><code class=\"ocaml\">let print_prime prime x =  \n  if x then print_int (prime()) (*called when needed*)\n  else print_int 0;;\n\n(*Create a thunk*)\nlet millionth_prime() = find_nth_prime 1000000;;\n\nprint_prime millionth_prime false;;  \n</code></pre>\n\n<p>Now, <code>print_prime</code> takes a thunk and we also create a capsule for the missionth prime. The thunk will be evaluated only before printing.</p>\n\n<h1 id=\"importantusages\">Important Usages</h1>\n\n<p>Thunk is the fundamental element for several classic usages:</p>\n\n<ul>\n<li><strong>stream_list</strong>: element is produced when being retrieved</li>\n<li><strong>lazy</strong>: evaluated only when needed and only once</li>\n<li><strong>async</strong>: concurrent computing, a kind of queue with lots scheduled computations. </li>\n</ul>\n\n<p>The above three and the important roles of thunks will be presented in details in future posts.</p>\n\n<hr>\n\n<p><em>To be continued</em></p>";;
              val s : string =
  "<p><img src=\"http://typeocaml.com/content/images/2014/11/thunk.jpg#hero\" alt=\"thunk\">\nA th…"
# let h = Cow.Html.of_string s;;
Exception: Parsing.Parse_error.
```

**Comment by Chris00** [comment]
Ideally, a parser inspired by `Nethtml` (which is quite robust) would be nice...

**Comment by rgrinberg** [comment]
Reusing some parser on opam should good as long as it doesn't bring any C or other massive dependencies IMO. Any other candidates?

**Comment by Drup** [comment]
I don't, unfortunately. The parser that is inside tyxml should not be used, and as soon as I can find something decent to replace it, I will (the current candidate is https://github.com/Eyyub/ppx_tyxml). This work is a bit on hold right now.

@dsheets did things on this subject (https://github.com/dsheets/htmlmu ?). 

Maybe it's time to fork some parts of ocamlnet ?

**Comment by Chris00** [comment]
I believe that the OCamlnet parser is a good starting choice but it should be updated to take care of [namespaces](http://www.w3.org/TR/xhtml1/#normative).

**Comment by Chris00** [comment]
(Then `Cow.Html.output` should by default deal properly with standard namespaces encountered in XHTML).

**Comment by dsheets** [comment]
#13 tracks the issue of simultaneously outputting valid XML and HTML.

I don't think the cow HTML parser is intended to be used for arbitrary HTML input. I have an unreleased tool that maps over URLs in (potentially sloppy) HTML and CSS to transform them in various ways used by ocaml.org. That tool is <https://github.com/dsheets/webmaster>. It uses ocamlnet to fallback from xmlm parsing to sloppy HTML parsing.

Is the example used in this issue failing due to HTML which is not XML or another problem? A reduced test case would be nice.

I'm not sure why cow should include a generic sloppy HTML parser as it doesn't seem to fit with the other components of the library which are mainly about generating code. Why not work on making a sloppy HTML parser into a standalone package?

**Comment by Drup** [comment]
I agree with the generic package proposal.

**Comment by Chris00** [comment]
OK. I suggest we then start a (X)HTML 5 parser/writer.  This is quite an endeavor though if one wants to be correct.

[See here](https://github.com/ocaml/ocaml.org/commits/cohttp) for the intended use (removing the non-portable ocamlnet dependency).

**Comment by dsheets** [comment]
It would be nice to have a sloppy HTML5 parser. I believe this library already implements an XHTML5 writer. Personally, I never want to output HTML that can't be parsed as XML...

You might be interested in <http://lists.w3.org/Archives/Public/public-html/2007Jul/1103.html>.

**Comment by aantron** [comment]
I think everyone in this thread is aware of [Markup.ml](https://github.com/aantron/markup.ml). It's not clear – does COW need to use a sloppy HTML5 parser?

If yes, I'd be happy to work on making Markup.ml suitable for COW. A polyglot writer should be easy to do.

**Comment by dsheets** [comment]
The polyglot HTML writer from `xmlm` data already exists in this library for some time and is now also a stand-alone package, [dsheets/polyglot](https://github.com/dsheets/polyglot), and has been released in the opam repository as `polyglot`. I don't think that most users of COW consume arbitrary input so I don't see why a sloppy parser should be a requirement to build the library.

I think a sloppy parser should either be optional or a FAQ/set of links to related libraries should be put in the README.

**Comment by dsheets** [comment]
Oh, it would also be nice for someone to integrate `polyglot` with COW. It should be straightforward as I just took my polyglot generation in COW and used/improved it to write `polyglot`.

**Comment by samoht** [comment]
No more parser, no more bug.
